### PR TITLE
CA-384537 support NBD for CD device

### PR DESCRIPTION
**Pull request by lindig**
Passing an NBD to qemu using qmp: open the socket and pass it to qmp.
Compared to the existing case that the CD is opened as a raw device the
handling of the socket is slightly different: it does not require a `Qmp.Remove_fd`.

### CODE REVIEW

**Comment by lindig** [comment]
This was manually tested with a build that uses NBD to provide the CD for guests.

File: ocaml/xenopsd/xc/device.ml [hunk 1]
```diff
@@ -3024,6 +3016,33 @@ module Backend = struct
           match params with
           | "" ->
               qmp_send_cmd domid Qmp.(Eject (cd, Some true)) |> ignore
+          | params when is_nbd params ->
+              let path, exportname = nbd params in
+              info "%s: domain=%d NBD socket=%s" __FUNCTION__ domid path ;
+              with_socket path @@ fun fd ->
+              let cmd = Qmp.(Add_fd None) in
```
Comment (edwintorok) [inline ocaml/xenopsd/xc/device.ml:3023]: This code could be shared with the non-nbd case by defining a function and giving it the path and whether to close the FD or not as parameters.

Unless we want to completely drop the non-nbd path eventually?

**Review by edwintorok — APPROVED** [review]
The changes themselves look ok

File: ocaml/xenopsd/xc/device.ml (hunk 1 above)
Comment (lindig) [reply ocaml/xenopsd/xc/device.ml:3023]: I expect that we will drop the non-nbd code eventually. If you look closely there are some difference between the two cases that are not so easy to factor out and so I am not convinced that a more complicated logic buys much.

**Comment by lindig** [comment]
Reviewers might want to review by commit because I simplified some error reporting code; the actual change is in one commit and the rest is improvements.

**Review by Vincent-lau — APPROVED** [review]
why is Remove_fd not needed in this special case?

**Comment by lindig** [comment]
@rosslagerwall might be able to explain. We initially had it and it didn't work. I suspect in the raw device case the FD is dup'ed but not in the socket case. It does feel inconsistent, though.

**Comment by rosslagerwall** [comment]
> @rosslagerwall might be able to explain. We initially had it and it didn't work. I suspect in the raw device case the FD is dup'ed but not in the socket case. It does feel inconsistent, though.

Indeed, this does appear to be what it is doing but I can't explain why it is inconsistent.

**Pull request merged**